### PR TITLE
Use tree-sitter-facade to enable Wasm support

### PR DESCRIPTION
**Pull request by torhovland**
First step on the way to implementing #171.

### CODE REVIEW

File: src/configuration.rs [hunk 1]
```diff
@@ -1,281 +1,80 @@
-use std::str::FromStr;
```
Comment (Xophmeister) [inline src/configuration.rs:1]: Reopens #311

File: src/error.rs [hunk 2]
```diff
@@ -101,7 +101,7 @@ impl Error for FormatterError {
     fn source(&self) -> Option<&(dyn Error + 'static)> {
         match self {
             Self::Idempotence => None,
-            Self::Internal(_, source) => source.as_ref().map(|e| e as &dyn Error),
+            Self::Internal(_, source) => source.as_ref().map(|e| &**e),
```
Comment (Xophmeister) [inline src/error.rs:104]: Isn't `&**e` equivalent to `*e`? Is this related to [this comment](https://tweag.slack.com/archives/CKMLS3094/p1678342865049809)?

File: src/language.rs [hunk 3]
```diff
@@ -193,5 +194,36 @@ impl From<&Language> for Vec<tree_sitter::Language> {
...
+            Language::Nickel => vec!["nickel"],
+            Language::Ocaml => vec!["ocaml", "ocaml-interface"],
+            Language::OcamlImplementation => vec!["ocaml"],
+            Language::OcamlInterface => vec!["ocaml-interface"],
+            Language::Rust => vec!["rust"],
+            Language::Toml => vec!["toml"],
+            Language::TreeSitterQuery => vec!["query"],
+        }
+        .iter()
+        .map(|name| format!("tree-sitter-{}.wasm", name))
```
Comment (Xophmeister) [inline src/language.rs:218]: These are presumably the grammar dependencies, compiled to WASM and distributed in the package (i.e., we can rely on this standard filename format)?

Also, what about `.mli`? IIRC, `Language::OCamlInterface` still uses `tree-sitter-ocaml`, but it exposes a separate `tree_sitter::Language`.

**tl;dr** If this works, then it may be worth documenting in comments why it works.

**Review by Xophmeister — APPROVED** [review]
I've left a few comments -- they're not really suggestions -- but LGTM :100:

File: .vscode/launch.json [hunk 4]
```diff
@@ -26,21 +26,22 @@
...
-                    "name": "tree-sitter-formatter",
+                    "name": "topiary",
                     "kind": "bin"
                 }
             },
             "args": [
-                "<", "tests/samples/input/rust.rs",
-                "-l", "rust"
+                "--query", "languages/debug.scm",
+                "--input-file", "languages/debug.scm",
```
[the remark this answers is not in the document]
Comment (torhovland) [reply .vscode/launch.json:43]: In my opinion, in tool specific config files like this, there are things that are user-specific settings, and other things that are useful setup config for everyone (or at least every user of the tool). So I'd say some of this makes sense to have checked in.

File: Cargo.toml [hunk 5]
```diff
@@ -11,25 +11,26 @@ clap = { version = "4.1", features = ["derive"] }
...
 log = "0.4"
-ouroboros = "0.15"
 pretty_assertions = "1.3"
+regex = "1.7.1"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 tempfile = "3.4.0"
 test-log = "0.2"
-tree-sitter = "0.20"
+tree-sitter-facade = { git = "https://github.com/tweag/tree-sitter-facade" }
```
[the remark this answers is not in the document]
Comment (torhovland) [reply Cargo.toml:20]: I don't think so. They are potentially useful outside of Topiary, for anyone who needs to target Tree-sitter both natively and through Wasm. We are clearly not the only ones who have run into that need.

And since https://github.com/silvanshade/tree-sitter-facade is archived, we are the de-facto maintainers of that code now.

File: src/configuration.rs (hunk 1 above)
Comment (torhovland) [reply src/configuration.rs:1]: Good point. I've done it.

File: src/error.rs (hunk 2 above)
Comment (torhovland) [reply src/error.rs:104]: `e` is an `&Box<dyn Error>`. `*e` results in `expected reference, found struct Box`. We need to deref that `Box` with an extra `*`. Now we have a `dyn Error`. But since the function wants to return an `&dyn Error` (wrapped in an `Option`), we add an `&`.

File: .vscode/launch.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (Xophmeister) [reply .vscode/launch.json:43]: Fair enough. My opinion on extra-repository stuff is much more strict, but it is still an opinion :) Given that, and the easiest thing to do here is nothing, then :+1:
